addons/stock_account/models/stock_move_line.py

```diff
@@ -48,7 +49,35 @@ def write(self, vals):
                 if float_is_zero(diff, precision_rounding=rounding):
                     continue
                 self._create_correction_svl(move, diff)
-        res = super(StockMoveLine, self).write(vals)
+        new_lot = False
+        if 'lot_id' in vals:
+            new_lot = vals.get('lot_id')
+        if 'quant_id' in vals:
+            new_quant = vals.get('quant_id')
+            new_lot = self.env['stock.quant'].browse(new_quant).lot_id.id
+        if new_lot:
+            # remove quantity of old lot
+            for move_line in self:
+                if move_line.state != 'done':
+                    continue
+                move = move_line.move_id
+                rounding = move.product_id.uom_id.rounding
+                diff = move.product_uom._compute_quantity(move_line.quantity, move.product_id.uom_id, rounding_method='HALF-UP')
+                if float_is_zero(diff, precision_rounding=rounding):
+                    continue
+                self._create_correction_svl(move, -diff)
```

Comment:
Not possible to introduce it in the upper loop and introduce the concept of lot in _create_correction_svl